modules/apps/callbacks/ibc_middleware.go

```diff
@@ -0,0 +1,337 @@
+	}
+
+	err = im.processCallback(ctx, packet, types.CallbackTypeReceivePacket, callbackData, callbackExecutor)
+	types.EmitCallbackEvent(ctx, packet, types.CallbackTypeReceivePacket, callbackData, err)
+
+	return ack
+}
+
+// WriteAcknowledgement implements the WriteAcknowledgement destination callbacks for the ibc-callbacks middleware
+// during asynchronous packet acknowledgement.
+// It defers to the underlying application and then calls the contract callback.
+// If the contract callback runs out of gas and may be retried with a higher gas limit then the state changes are
+// reverted via a panic.
+func (im IBCMiddleware) WriteAcknowledgement(
+	ctx sdk.Context,
+	chanCap *capabilitytypes.Capability,
+	packet ibcexported.PacketI,
+	ack ibcexported.Acknowledgement,
+) error {
+	err := im.ics4Wrapper.WriteAcknowledgement(ctx, chanCap, packet, ack)
+	if err != nil {
+		return err
+	}
+
+	callbackData, err := types.GetDestCallbackData(im.app, packet, ctx.GasMeter().GasRemaining(), im.maxCallbackGas)
+	if err != nil {
+		return nil
+	}
+
+	callbackExecutor := func(cachedCtx sdk.Context) error {
+		return im.contractKeeper.IBCReceivePacketCallback(cachedCtx, packet, ack, callbackData.CallbackAddress)
+	}
+
+	err = im.processCallback(ctx, packet, types.CallbackTypeReceivePacket, callbackData, callbackExecutor)
+	types.EmitCallbackEvent(ctx, packet, types.CallbackTypeReceivePacket, callbackData, err)
+
+	return nil
+}
+
+// processCallback executes the callbackExecutor and reverts contract changes if the callbackExecutor fails.
+//
+// panics if
+//   - the callbackType is SendPacket and the contractExecutor panics for any reason, or
+//   - the contractExecutor out of gas panics and the relayer has not reserved gas grater than or equal to
+//     CommitGasLimit.
+func (IBCMiddleware) processCallback(
+	ctx sdk.Context, packet ibcexported.PacketI, callbackType types.CallbackType,
+	callbackData types.CallbackData, callbackExecutor func(sdk.Context) error,
+) (err error) {
+	cachedCtx, writeFn := ctx.CacheContext()
+	cachedCtx = cachedCtx.WithGasMeter(sdk.NewGasMeter(callbackData.ExecutionGasLimit))
+
+	defer func() {
+		ctx.GasMeter().ConsumeGas(cachedCtx.GasMeter().GasConsumedToLimit(), fmt.Sprintf("ibc %s callback", callbackType))
+		if r := recover(); r != nil {
+			// We handle panic here. This is to ensure that the state changes are reverted
+			// and out of gas panics are handled.
+			//
+			// We do not handle panics for SendPacket callbacks because we require an approval
+			// from the callback actor to send the packet.
+			if callbackType == types.CallbackTypeSendPacket {
+				panic(r)
+			}
+
+			if oogError, ok := r.(sdk.ErrorOutOfGas); ok {
+				// If execution gas limit was less than the commit gas limit, allow retry.
+				if callbackData.ExecutionGasLimit < callbackData.CommitGasLimit {
+					panic(r)
+				}
+				types.LogDebugWithPacket(ctx, callbackType, packet, "Callbacks recovered from out of gas panic.", "panic", oogError)
+			}
```

Comment:
I think probably the behavior we want is:
```go
if r := recover(); r != nil {
    // A callback must be skipped on non SendPacket callback types when the callbackData
    // is not allowed to be retried for execution
    if callbackType != types.CallbackTypeSendPacket && !callbackData.AllowRetry() {
        return // do not propagate panic, force transaction to complete 
    }
    panic(r) // propagate original panic

}
```
These comments could be nicer, but the general idea is there are a lot of cases we want to keep the panic and only a single case where we catch the panic and return nil

Reply:
Yep, nice refactor!

Reply:
Yea, that's right. The case we ignore the panic is if all the following is true:
* callback type is not SendPacket
* panic is out of gas
* callbackData is allowed to be retried

So something like:
```go
if r := recover(); r != nil {
            // The provided panic will only be dropped if the panic is due to an out of gas error, 
            // the current callback is not SendPacket, and the callback is not allowed to be retried.
            if oogError, ok := r.(sdk.ErrorOutOfGas); ok {
                // If execution gas limit was less than the commit gas limit, allow retry.
                if callbackType != types.CallbackTypeSendPacket && !callbackData.AllowRetry() {
                    return // commit tx
                }
            }
            panic(r)
        }
```

Reply:
I'll commit this regardless as this is more concise and comments do a reasonable job imo